source/ports/rs_port/src/lib.rs

```diff
@@ -18,288 +18,303 @@
-						let st = CString::new(x.clone()).expect("can't convert to c str");
-						let a = metacall_value_create_string(st.as_ptr(), x.len());
-						c_args.push(a);
-					},
-					_ => {}
-				}
-			}
-			let ret : *mut c_void = metacallfv(c_func, c_args.as_mut_ptr());
-			let mut rt = Any::Null;
-			if !ret.is_null() {
-				/* TODO: This should be done by an enum or something mimicking the enum in metacall.h */
-				match metacall_value_id(ret) {
-					0 => {
-						rt = Any::Bool(if metacall_value_to_bool(ret) != 0 { true } else { false });
-					},
-					1 => {
-						rt = Any::Char(metacall_value_to_char(ret) as u8 as char);
-					},
-					2 => {
-						rt = Any::Short(metacall_value_to_short(ret));
-					},
-					3 => {
-						rt = Any::Int(metacall_value_to_int(ret));
-					},
-					4 => {
-						rt = Any::Long(metacall_value_to_long(ret));
-					},
-					5 => {
-						rt = Any::Float(metacall_value_to_float(ret));
-					},
-					6 => {
-						rt = Any::Double(metacall_value_to_double(ret));
-					},
-					7 => {
-						let st = std::ffi::CStr::from_ptr(metacall_value_to_string(ret));
-						rt = Any::Str(String::from(st.to_str().expect("couldn't convert CStr to &str")));
-					},
-					8 => {
-						// METACALL_BUFFER
-					},
-					9 => {
-						// METACALL_ARRAY
-					},
-					10 => {
-						// METACALL_MAP
-					},
-					11 => {
-						// METACALL_PTR
-					},
-					12 => {
-						// METACALL_FUTURE
-					},
-					13 => {
-						// METACALL_FUNCTION
-					},
-					14 => {
-						rt = Any::Null;
-					},
-					_ => {}
-				}
-				metacall_value_destroy(ret);
-			}
-			for arg in c_args {
-				metacall_value_destroy(arg);
-			}
-			return Ok(rt);
-		}
-	}
+// Possible types as variants in Rust
+pub fn metacall<'a>(
+    func: &str,
+    args: impl IntoIterator<Item = &'a Any>,
+) -> Result<Any, &'static str> {
+    let c_function = CString::new(func).expect("Conversion to C String failed");
+    unsafe {
+        // let c_func = metacall_function(c_function.as_ptr());
+        let c_func: *mut c_void = metacall_function(c_function.as_ptr());
+        if c_func.is_null() {
+            return Err("Function Not Found");
+        }
+        let mut c_args: Vec<*mut c_void> = args
+            .into_iter()
+            .map(|arg| match arg {
+                Any::Short(x) => metacall_value_create_short(*x),
+                Any::Int(x) => metacall_value_create_int(*x),
+                Any::Long(x) => metacall_value_create_long(*x),
+                Any::Float(x) => metacall_value_create_float(*x),
+                Any::Double(x) => metacall_value_create_double(*x),
+                Any::Bool(x) => metacall_value_create_bool(*x as c_int),
+                Any::Char(x) => metacall_value_create_char(*x as c_char),
+                Any::Str(x) => {
+                    let st = CString::new(x.as_str()).expect("can't convert to c str");
+                    metacall_value_create_string(st.as_ptr(), x.len())
+                }
```

Comment:
The documentation isn't clear on whether `metacall_value_to_string` requires a null-terminated string, as the size is passed along with it. If it isn't required, then

```suggestion
                Any::Str(x) => metacall_value_create_string(x.as_ptr() as *const _, x.len()),
```

Reply:
I have added more doc about that: https://github.com/metacall/core/commit/51b85c876df753f1b89738cc253fefb9bacd9984

I also realized that metacall_value_from_string is very dangerous if not used properly, it truncates the size if the size is smaller (without leaving null terminated string), or it leads a gap in the end if the size is bigger (but in this case it is null terminated). I don't know what implications can have this, but when using the API to implement the Ports, I try to copy values when possible and I threat it as what it is, a C API.